Add support for background and session-based apps (#2651)
beeware/toga: docs/reference/api/app.rst

This approach to app construction is most useful with simple apps. For most complex
apps, you should subclass :class:`toga.App`, and provide an implementation of
:meth:`~toga.App.startup()`. This implementation *must* assign a value to
:attr:`~toga.App.main_window` for the app. The value that is assigned controls the type
of app that is created. The possible values that can be assigned to
:attr:`~toga.App.main_window` is :ref:`discussed below <assigning-main-window>`; the

> most common type of app will assign an instance of :any:`toga.MainWindow`:

Reviewer: Still suggest removing the sentence about "type of app", especially since the `main_window` can be changed dynamically.

This approach to app construction is most useful with simple apps. For most complex
apps, you should subclass :class:`toga.App`, and provide an implementation of
:meth:`~toga.App.startup()`. This implementation *must* assign a value to

> :attr:`~toga.App.main_window` for the app. The possible values are :ref:`discussed 
> below <assigning-main-window>`; most apps will assign an instance of 

Reviewer: ```suggestion
:attr:`~toga.App.main_window` for the app. The possible values are :ref:`discussed
below <assigning-main-window>`; most apps will assign an instance of
```

you wish to customize the menu items exposed by your app, you can add or remove commands
in your :meth:`~toga.App.startup()` implementation.

.. _assigning-main-window:

Assigning a main window
-----------------------

An app *must* assign ``main_window`` as part of the startup process. However, the value
that is assigned as the main window will affect the behavior of the app.

:class:`toga.Window`

> ~~~~~~~~~~~~~~~~~~~~

Author: Agreed; however, AIUI, the `~` is the preferred way to do this; that keeps the Sphinx reference explicit, while hiding the context.

> ~~~~~~~~~~~~~~~~~~~~

Reviewer: Another advantage of using `:any:` is that it gives an error if it doesn't find a valid target, while `:class:`, `:meth:`, etc. all fail silently and output the text with no hyperlink.

the Finder to open a second document of a type managed by the app, it will be opened in
the existing app instance. Closing all windows will not cause the app to exit; the app
will keep executing until explicitly exited.

If the app is started without an explicit file reference (e.g., by specifying a filename
at the command line, or dragging a file onto the app's icon), a file dialog will be
displayed prompting the user to select a file to open. If this dialog is dismissed, the
app will continue running. Selecting "Open" from the file menu will also display this
dialog. If a file is selected, a new document window will be opened.

Linux/Windows
^^^^^^^^^^^^^

On Linux and Windows, a single app instance app can also manage multiple windows;
however when the last window being managed by an app instance is closed, the app
instance will exit. If the App is started without an explicit file reference, an empty
document of the default document type will be opened.

Mobile, web and console
^^^^^^^^^^^^^^^^^^^^^^^

Session-based apps are not supported on mobile, web and console platforms.

Background app
~~~~~~~~~~~~~~

> To create an app without *any* main window, assign ``toga.BACKGROUND`` as the main
> window. This will allow your app to persist even if it doesn't have any open windows. It

Reviewer: `None` also means "an app without any main window".

```suggestion
If you set :attr:`~toga.App.main_window` to :attr:`App.BACKGROUND`, this will
allow your app to persist even if it doesn't have any open windows. It
```
